Tests/SourceKitTests/BuildSystemTests.swift

```diff
@@ -0,0 +1,167 @@
+    testServer.server!.workspace = workspace
+
+    sk = testServer.client
+    _ = try! sk.sendSync(InitializeRequest(
+        processId: nil,
+        rootPath: nil,
+        rootURL: nil,
+        initializationOptions: nil,
+        capabilities: ClientCapabilities(workspace: nil, textDocument: nil),
+        trace: .off,
+        workspaceFolders: nil))
+  }
+
+  override func tearDown() {
+    buildSystem = nil
+    workspace = nil
+    sk = nil
+    testServer = nil
+  }
+
+  func testDocumentUpdatedBuildSettings() {
+    let url = URL(fileURLWithPath: "/a.swift")
+
+    sk.sendNoteSync(DidOpenTextDocument(textDocument: TextDocumentItem(
+      url: url,
+      language: .swift,
+      version: 12,
+      text: """
+      func
+      """
+    )), { (note: Notification<PublishDiagnostics>) in
+      XCTAssertEqual(note.params.diagnostics.count, 1)
+      XCTAssertEqual("func", self.workspace.documentManager.latestSnapshot(url)!.text)
+    })
+
+    // Modify the build settings and inform the SourceKitServer.
+    // This should trigger new diagnostics.
+    buildSystem.buildSettingsByFile[url] = FileBuildSettings(compilerArguments: [])
+    testServer.server?.fileBuildSettingsChanged([url])
+
+    let expectation = XCTestExpectation(description: "refresh")
+    sk.handleNextNotification { (note: Notification<PublishDiagnostics>) in
+      XCTAssertEqual(note.params.diagnostics.count, 1)
+      XCTAssertEqual("func", self.workspace.documentManager.latestSnapshot(url)!.text)
+      expectation.fulfill()
+    }
+
+    let result = XCTWaiter.wait(for: [expectation], timeout: 5)
+    if result != .completed {
+      fatalError("error \(result) waiting for diagnostics notification")
+    }
+  }
+
+  func testDocumentBuildSettingsChangedFalseAlarm() {
+    let url = URL(fileURLWithPath: "/a.swift")
+
+    sk.sendNoteSync(DidOpenTextDocument(textDocument: TextDocumentItem(
+      url: url,
+      language: .swift,
+      version: 12,
+      text: """
+      func
+      """
+    )), { (note: Notification<PublishDiagnostics>) in
+      XCTAssertEqual(note.params.diagnostics.count, 1)
+      XCTAssertEqual("func", self.workspace.documentManager.latestSnapshot(url)!.text)
+    })
+
+    // Modify the build settings and inform the SourceKitServer.
+    // This shouldn't trigger new diagnostics since nothing actually changed (false alarm).
+    testServer.server?.fileBuildSettingsChanged([url])
+
+    let expectation = XCTestExpectation(description: "refresh doesn't occur")
+    expectation.isInverted = true
+    sk.handleNextNotification { (note: Notification<PublishDiagnostics>) in
+      XCTAssertEqual(note.params.diagnostics.count, 1)
+      XCTAssertEqual("func", self.workspace.documentManager.latestSnapshot(url)!.text)
+      expectation.fulfill()
+    }
+
+    let result = XCTWaiter.wait(for: [expectation], timeout: 5)
+    if result != .completed {
+      fatalError("error \(result) waiting for diagnostics notification")
+    }
+  }
+}
```

Comment:
How hard would it be to add a test for clangd as well?

Reply:
Hmm, I can try, just might also need to do something like the FallbackBuildSystem to have some proper flags

Reply:
Should I just add them to the `LocalClangTests`? Not sure how the `requireClangd` behavior works on CI

Reply:
You can ignore `requireClangd`, which is a debugging aid.  It's `haveClangd` that matters.  You can just copy the trivial definition:

```
haveClangd = ToolchainRegistry.shared.toolchains.contains { $0.clangd != nil }

func testFoo() {
  guard haveClangd else { return }
  ...
}
```

None of this should matter in CI though, since all our bots should have clangd at this point.  This was mostly to make it possible to test locally if you don't have clang

Reply:
Thanks, added